Mathlib/CategoryTheory/Subpresheaf/Finite.lean

```diff
@@ -0,0 +1,171 @@
+/-
+Copyright (c) 2025 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Subpresheaf.OfSection
+
+/-!
+# Subpresheaves that are generated by finitely many sections
+
+Given `F : Cᵒᵖ ⥤ Type w`, `G : Subpresheaf F`, objects `X : ι → Cᵒᵖ` and sections
+`x : (i : ι) → F.obj (X i)`, we define a predicate `G.IsGeneratedBy x` saying
+that `G` is the subpresheaf generated by the sections `x i`. If this holds for
+a finite index type `ι`, we say that `G` is "finite", and this gives a type
+class `G.IsFinite`.
+
+-/
+
+universe w'' w' w v u
+
+namespace CategoryTheory
+
+open Opposite
+
+variable {C : Type u} [Category.{v} C] {F : Cᵒᵖ ⥤ Type w}
+
+namespace Subpresheaf
+
+variable (G : Subpresheaf F)
+
+section
+
+variable {ι : Type w'} {X : ι → Cᵒᵖ} (x : (i : ι) → F.obj (X i))
+
+/-- A subpresheaf `G : Subpresheaf F` is generated by sections `x i : F.obj (X i)`
+if `⨆ (i : ι), ofSection (x i) = G`. -/
+def IsGeneratedBy : Prop := ⨆ (i : ι), ofSection (x i) = G
+
+lemma isGeneratedBy_iff : G.IsGeneratedBy x ↔ ⨆ (i : ι), ofSection (x i) = G := Iff.rfl
+
+namespace IsGeneratedBy
+
+variable {G x} (h : G.IsGeneratedBy x)
+include h
+
+lemma iSup_eq : ⨆ (i : ι), ofSection (x i) = G := h
+
+lemma ofSection_le (i : ι) : ofSection (x i) ≤ G := by
+  rw [← h.iSup_eq]
+  exact le_iSup (fun i ↦ ofSection (x i)) i
+
+lemma mem (i : ι) : x i ∈ G.obj (X i) := by
+  rw [← ofSection_le_iff]
+  exact h.ofSection_le i
+
+lemma of_equiv {ι' : Type w''} (e : ι' ≃ ι) :
+    G.IsGeneratedBy (fun i' ↦ x (e i')) := by
+  rw [isGeneratedBy_iff, ← h.iSup_eq]
+  apply le_antisymm
+  · simp only [iSup_le_iff]
+    intro i'
+    exact le_iSup (fun i ↦ ofSection (x i)) (e i')
+  · simp only [iSup_le_iff]
+    intro i'
+    convert (le_iSup (fun i' ↦ ofSection (x (e i'))) (e.symm i')) <;> simp
+
+lemma image {F' : Cᵒᵖ ⥤ Type w} (f : F ⟶ F') :
+    (G.image f).IsGeneratedBy (fun i ↦ f.app _ (x i)) := by
+  simp only [isGeneratedBy_iff, ← h.iSup_eq, image_iSup, ofSection_image]
+
+end IsGeneratedBy
+
+end
+
+/-- A subpresheaf of types is "finite" if it is generated by finitely many sections. -/
+class IsFinite : Prop where
+  exists_isGeneratedBy :
+    ∃ (ι : Type) (_ : Finite ι) (X : ι → Cᵒᵖ) (x : (i : ι) → F.obj (X i)),
+      Nonempty (G.IsGeneratedBy x)
+
+namespace IsFinite
+
+variable [hG : G.IsFinite]
+
+/-- A choice of index type for the generating sections of a finitely generated subpresheaf. -/
+def index : Type := hG.exists_isGeneratedBy.choose
```

Comment:
This is a type, so it should start with a capital letter.

Reply:
Thanks!